### DISCUSSION TITLE
Inconsistency in inheritance structure and make_node

### DISCUSSION THREAD

**Discussion by ferrine** [pymc-devs/pytensor#148] [General]
The `make_node` in the base [Op](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/graph/op.py#L148) class is meant to deal with class [Variable](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/graph/basic.py#L318) which in turn has a base [Type](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/graph/type.py#L14) class. In a number of places we violate the principle of common interface for this type structure. 

1. There are numerous as_tensor_variable in the codebase, just scroll down [this](https://github.com/search?q=repo%3Apymc-devs%2Fpytensor%20make_node&type=code)
2. Implement specific checks for inferring the just right type in the make_node https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/tensor/random/basic.py#L1914
3. we make [redundant](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/breakpoint.py#L74) checks according to the signature we have
4. we also assume [more](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/breakpoint.py#L79) about the type than it should be (here the above isinstance should actually be TensorType) because Variable does not have any ndim, that is a property of [TensorType](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/tensor/type.py#L382) and could actually be a property from [HasShape](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/graph/type.py#L271)
5. [HasDytpe](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/graph/type.py#L268) looks also suspicious and violates type structure, it declares `dtype: str` later on every sublcass picks up this dtype, compares to numpy dtype and creates an attribute with numpy_dtype (there are just three places HasDType is Inherited)
    1. https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/scalar/basic.py#L272
    2. https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/tensor/type.py#L58
    3. https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/sparse/type.py#L36 (it even inherits the HasDtype twice)
    4. the function dtype_spec is defined outside of HasDtype in every inherited class
        1. https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/scalar/basic.py#L373
        2. https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/tensor/type.py#L279
        3. https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/tensor/type.py#L40
        4. https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/sparse/type.py#L51
    5. I guess all the dtype specific things utterly have to be in a single consolidated space, the HasDtype class or HasNumpyDtype as every single class uses this property to ensure numpy dtype is present there. Inheritance means more than having a dtype string and similar checks are done in every place.
    6. TensorType seems to be the entry point of classes that both have dtype and shape and is the most generic, however, seemingly more specific class [ScalarType](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/scalar/basic.py#L272) is not a subclass of the TensorType which seems weird `issubclass(ScalarType, TensorType) is False`, but in fact, TensorType is a superset of the TensorType
    7. The shape TensorType seems to be uniform across all the codebase and is a subset of TensorType which is [lvector with fixed length](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/tensor/shape.py#L72) Creating Shape(TensorType) makes total sense since more structure is asserted in this instance: fixed dtype + fixed shape
        1. there is also inconsistency with input types, a generic variable is [considered](https://github.com/pymc-devs/pytensor/blob/d7985536ae8674ab9af62b964fd6c9ff566ee3de/pytensor/tensor/shape.py#L67) as input but only `HasShape` typed classes are valid for taking shape, isn't it?

**Comment by ricardoV94** [comment]
About types, this might be relevant: https://github.com/aesara-devs/aesara/pull/1207

And the associated issue (and its "off-topic" comments): https://github.com/aesara-devs/aesara/issues/134

**Reply by ferrine** [reply]
yeah I was thinking of similar things
